### ISSUE TITLE
LWC does not convert pre-existing attributes to props during custom element upgrade

### ISSUE THREAD

**Issue by nolanlawson** [salesforce/lwc#3478]
### Description

Pre-existing attributes on an HTML element, when combined with `CustomElementConstructor`, do not result in those attributes being converted into props set on the LWC component. As described [in this post](https://nolanlawson.com/2021/08/03/handling-properties-in-custom-element-upgrades/), we should probably convert attributes to props in the `constructor` (as Lit does).

```html
<x-app foo="foo"></x-app>
```

```js
import App from "x/app";

// the `foo` attribute is ignored here
customElements.define('x-app', App.CustomElementConstructor)
```

#### Steps to Reproduce

[Repro](https://stackblitz.com/edit/salesforce-lwc-tdncg7?file=src/main.js)

**Comment by jayanth-kumar-morem** [comment]
A possible solution to this problem is to handle the conversion of attributes to props within the `connectedCallback` lifecycle hook of your LWC component. This way, when your component is connected to the DOM, it can read the attributes and set the corresponding props.

Here's an example of how you can achieve this in your `x/app` component:

```
import { LightningElement, api } from 'lwc';

export default class App extends LightningElement {
    _foo;

    // Create a getter and setter for the foo property in your x/app component:
    @api
    get foo() {
        return this._foo;
    }

    set foo(value) {
        this._foo = value;
    }

    // Implement the connectedCallback lifecycle hook in your x/app component to convert the foo attribute to a prop:
    connectedCallback() {
        if (this.hasAttribute('foo')) {
            this.foo = this.getAttribute('foo');
        }
    }
}

```

**Comment by yashpatel1998** [comment]
@nolanlawson Can I pick this up?

**Comment by nolanlawson** [comment]
@yashpatel1998 Yes, but I think the bug description is actually wrong. Let me fix it.

**Comment by yashpatel1998** [comment]
@nolanlawson  `attributeChangedCallback` does handle pre-existing attributes provided they are public and marked as  `@api`. 
As per my understanding of the issue, we would like to set non public props on a `CustomElementConstructor` if they are passed as attributes. Am I correct?

**Comment by nolanlawson** [comment]
@yashpatel1998 You're right ([repro](https://stackblitz.com/edit/salesforce-lwc-hj3gu2?file=index.html,src%2Fmain.js,src%2Fmodules%2Fx%2Fapp%2Fapp.js,src%2Fmodules%2Fx%2Fapp%2Fapp.html)). I must have forgotten to add `@api`. 😅 

I'll update the bug title and description.

**Comment by nolanlawson** [comment]
> As per my understanding of the issue, we would like to set non public props on a CustomElementConstructor if they are passed as attributes. Am I correct?

@yashpatel1998 Sorry, this is not correct. I've updated the bug description with the expected behavior.

It's not about `@api` – it's about setting a prop on an HTMLElement before `customElements.define()` is called. Pre-existing _props_ (not attributes) are not handled correctly.

**Comment by yashpatel1998** [comment]
@nolanlawson Ah, this description is clear to me now. 
Although the prop is set before the `CustomElementConstructor` invocation the compiler does not set it while upgrading it via `customElements.define()`
